app/shared/app-data/src/androidMain/kotlin/domain/torrent/service/ServiceConnectionManager.kt

```diff
@@ -0,0 +1,211 @@
+                cont.resume(false)
+            } else {
+                logger.debug { "Service started, component name: $result" }
+            }
+        }
+        if (!startResult) {
+            return null
+        }
+
+        val currentDeferred = binderDeferred.value
+        if (!currentDeferred.isCompleted) {
+            currentDeferred.cancel()
+        }
+        val newDeferred = CompletableDeferred<T?>()
+        binderDeferred.value = newDeferred
+
+        val bindResult = context.bindService(
+            Intent(context, AniTorrentService.actualServiceClass),
+            this@ServiceConnectionManager,
+            Context.BIND_ABOVE_CLIENT,
+        )
+        if (!bindResult) return null
+
+        return try {
+            newDeferred.await()
+        } catch (ex: CancellationException) {
+            // onServiceDisconnected will cancel the deferred
+            null
+        }
+    }
+
+    override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
+        if (service == null) {
+            logger.error { "Service is connected, but got null binder!" }
+        }
+        binderDeferred.value.complete(mapBinder(service))
+    }
+
+    override fun onServiceDisconnected(name: ComponentName?) {
+        binderDeferred.value.cancel(CancellationException("Service disconnected."))
+        _connection.onServiceDisconnected()
+    }
+}
+
+private class ForegroundServiceTimeLimitObserver(
+    private val context: Context,
+    onServiceTimeLimitExceeded: () -> Unit
+) : DefaultLifecycleObserver {
+    private val logger = logger<ForegroundServiceTimeLimitObserver>()
+
+    private val acquireWakeLockIntent = Intent(context, AniTorrentService.actualServiceClass).apply {
+        putExtra("acquireWakeLock", 1.minutes.inWholeMilliseconds)
+    }
+    private var registered = false
+    private val timeExceedLimitIntentFilter = IntentFilter(AniTorrentService.INTENT_BACKGROUND_TIMEOUT)
+    private val timeExceedLimitReceiver = object : BroadcastReceiver() {
+        override fun onReceive(context: Context?, intent: Intent?) {
+            onServiceTimeLimitExceeded()
+        }
+    }
+
+    @RequiresApi(31)
+    override fun onPause(owner: LifecycleOwner) {
+        super.onPause(owner)
+
+        // app 到后台的时候注册监听
+        if (!registered) {
+            ContextCompat.registerReceiver(
+                context,
+                timeExceedLimitReceiver,
+                timeExceedLimitIntentFilter,
+                ContextCompat.RECEIVER_NOT_EXPORTED,
+            )
+            registered = true
+        }
+        try {
+            // 请求 wake lock, 如果在 app 中息屏可以保证 service 正常跑 [acquireWakeLockIntent] 分钟.
+            context.startService(acquireWakeLockIntent)
+        } catch (ex: ForegroundServiceStartNotAllowedException) {
+            // 大概率是 ServiceStartForegroundException, 服务已经终止了, 不需要再请求 wakelock.
+            logger.warn(ex) { "Failed to acquire wake lock. Service has already died." }
```

Comment:
```suggestion
            logger.warn { "Failed to acquire wake lock due to ForegroundServiceStartNotAllowedException" }
```
message 没有表达真正的原因 ("Service died" 原因不明)